pkg/store/metrics/metrics.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package metrics
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+)
+
+// seriesStatsAggregator aggregates results from fanned-out queries into a histogram given their
+// response's shape.
+type seriesStatsAggregator struct {
+	queryDuration *prometheus.HistogramVec
+
+	seriesLeBuckets  []int64
+	samplesLeBuckets []int64
+	seriesStats      storepb.SeriesStatsCounter
+}
+
+// NewSeriesStatsAggregator is a constructor for seriesStatsAggregator.
+func NewSeriesStatsAggregator(
+	reg prometheus.Registerer,
+	durationQuantiles []float64,
+	sampleQuantiles []int64,
+	seriesQuantiles []int64,
+) *seriesStatsAggregator {
+	return &seriesStatsAggregator{
+		queryDuration: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "thanos_store_api_query_duration_seconds",
+			Help:    "Duration of the Thanos Store API select phase for a query.",
+			Buckets: durationQuantiles,
+		}, []string{"series_le", "samples_le"}),
+		seriesLeBuckets:  seriesQuantiles,
+		samplesLeBuckets: sampleQuantiles,
+		seriesStats:      storepb.SeriesStatsCounter{},
+	}
+}
+
+// Aggregate is an aggregator for merging `storepb.SeriesStatsCounter` for each incoming fanned out query.
+func (s *seriesStatsAggregator) Aggregate(stats storepb.SeriesStatsCounter) {
+	s.seriesStats.Series += stats.Series
+	s.seriesStats.Samples += stats.Samples
+	s.seriesStats.Chunks += stats.Chunks
+}
+
+// Observe commits the aggregated SeriesStatsCounter as an observation.
+func (s *seriesStatsAggregator) Observe(duration float64) {
+	if s.seriesStats.Series == 0 || s.seriesStats.Samples == 0 || s.seriesStats.Chunks == 0 {
+		return
+	}
+	// Bucket matching for series/labels matchSeriesBucket/matchSamplesBucket => float64, float64
+	seriesLeBucket := s.findBucket(float64(s.seriesStats.Series), s.seriesLeBuckets)
+	samplesLeBucket := s.findBucket(float64(s.seriesStats.Samples), s.samplesLeBuckets)
+	s.queryDuration.With(prometheus.Labels{
+		"series_le":  strconv.Itoa(int(seriesLeBucket)),
+		"samples_le": strconv.Itoa(int(samplesLeBucket)),
```

Comment:
I know we agreed in proposal to this, but can we double check how to use this metric? How to visualizate this in Grafana/Prometheus? Can we make short experiment. I think it's quite a precedence here - I don't think `histogram_quantile` would work with this 🤔

Reply:
We can visualize it with queries like so:

```
histogram_quantile(0.95,
    sum(rate(
        thanos_store_api_query_duration_seconds{samples_le=1000, series_le=10}[5m]
    )) by (le)
)
```

As long as we don't mix different histograms, as in `histogram_quantile(0.95, sum(rate(thanos_store_api_query_duration_seconds{samples_le=~"1000|10000", series_le=~"10|100"}[5m])) by (le))`, we won't have issues with mixed up bucketing and the queries will be fine.

Reply:
Actually it would work fine even on `histogram_quantile(0.95, sum(rate(thanos_store_api_query_duration_seconds{samples_le=~"1000|10000", series_le=~"10|100"}[5m])) by (le))`, because the buckets are the same. 

We can aggregate visualize it analogous to this example from Prometheus' documentation: `histogram_quantile(0.9, sum by (job, le) (rate(http_request_duration_seconds_bucket[10m])))`:

```
histogram_quantile(0.95,
    sum(rate(
        thanos_store_api_query_duration_seconds{samples_le=1000, series_le=10}[5m]
    )) by (samples_le, series_le, le)
)
```

Reply:
> I don't think histogram_quantile would work with this 🤔

@bwplotka why do you think so? I am not able to see why it wouldn't work, but I don't have as much experience as you on this.